### ISSUE TITLE
use of new without delete

### ISSUE THREAD

**Issue by jackokring** [LMMS/lmms#2475]
looking at some of the code for instruments, there is often (SID for example) the use of *new* without *delete*. This is a memory hole, maybe not a serious one, but none the less (C++ does not collect pointer referenced objects, it is not Java). I also noted that sometimes when *delete* is called, it is called before rather than safely after detaching, from use by enclosing objects.

**Comment by tresf** [comment]
@jackokring Thanks for pointing this out (we need all the help we can get), but be careful not to open a "this common mistake all over the code shouldn't have been made" bug reports, as they'll never get closed.

Instead, we can put this on our futuremap of things to do (something that hasn't been actively maintained, but is out there and open), or alternately you can make a placeholder for cleaning up `new` resources on plugins x, y and z.

If a goal is defined, it makes it much easier to mark it as completed, and as you can see from our tracker, we have far too many unclear, not yet started and incomplete bug reports. :+1:

**Comment by jackokring** [comment]
@tresf I never put something on the futuremap. How is this done?

**Comment by tresf** [comment]
#908 is where it's been sitting, but it's a bit dusty.  How about you define the task better and we leave it open?

**Comment by jackokring** [comment]
The defined task: For each plugin obtain a list of *new* allocations of objects. These are the ones that use heap memory, and can lead to memory holes in the long term. In the class destructors, *delete* has to be used to free these heap allocations.

Notes:

1. If other things are done in the destructor, place the *delete(s)* as the last code in the destructor. This may be essential to provide thread safe code, by preventing reallocation on the heap before all possible variable writes have completed.

2. If the *new* is not done directly in a constructor, then the *delete* should be placed as the last statement to be done before the pointer to the *new* class falls out of the variable scope. This kind of code is rare, and indicates that there may also be the option of keeping the *new* class in scope in the class by adding a class variable, and *delete* in the destructor.

3. If it is not known at compile time how many *new* classes may be instanced, it is perhaps better to keep a pointer chain (linked list) of class instances, and *delete* them all in a while loop in the destructor. This keeps deallocation code efficient. Although finding a solution to fix the maximum number of allocations, and allocating an array would be advantageous for code execution speed.

4. It is more important to fix non *delete* memory issues in classes which are instanced via *new* often, than it is to fix the same problem in classes which are instanced just once, or statically instanced. A one off memory drip is less problematic than a continuous leak.

Sorry for such a general description as all (or most) plugins, but the advice to copy a plugin as a basis for starting to write your own plugin has perhaps lead to copying memory leak issues.

**Comment by michaelgregorius** [comment]
@jackokring Have you checked whether there really is a problem, i.e. have you run LMMS in valgrind and it has reported actual leaks from the plugins? If not I would add doing so as step 0.

After all it might also be that the memory allocation scheme is not that straight forward, e.g. that some memory is allocated in one class but deallocated by another. In such cases there would be no leak at all.

Anyway, thanks for your help! As @tresf has pointed out LMMS can use every helping hand.

**Comment by tresf** [comment]
@jackokring Thanks again for the detailed writeup.  Since this is a code quality issue, we should track this on the futuremap (not actively maintained, we need help).

If you can answer @michaelgregorius's questions or followup with the work you planned on, we can keep this open under better definition.  If not, we'll close it out. :+1:

**Comment by Wallacoloo** [comment]
Also, is there a reason we should be using `delete` instead of the newer C++ smart pointers (like `std::unique_ptr`, etc)? The way I see it, using these wrappers makes ownership and ownership transfers (e.g. when memory is allocated in one class and then deallocated in another, as @michaelgregorius mentions) more explicit, and thus less error-prone and easier to follow.

**Comment by michaelgregorius** [comment]
@Wallacoloo I agree that in some places it might be better to use the new smart pointers especially if the allocation and deallocation is done in a rather narrow scope. When using `std::shared_ptrs` one has to make sure that they are not passed around in a rather uncontrolled fashion as this can also create problems (cyclic pointing which in turn leads to memory holes).

With regards to this specific issue I propose that we close it for the following reasons:
* As @tresf has pointed out it is not very specific.
* The missing `delete` in the original post of @jackokring is not a real problem because the `voiceObject` that's allocated inherits from `QObject` and it is in fact deleted when the parent `sidInstrument` is deleted. I have checked this by putting break points in the corresponding destructors.
* I have checked with Valgrind's memory checking tool and found no leaks with regards to the instrument plugins. I have added and removed each instrument once (except the VST stuff which I do not build on my main machine) and no memory holes related to the LMMS code were reported. There's one hole reported for the `sf2_player` but it seems to be caused by code in the used library.

**Comment by softrabbit** [comment]
>  I have added and removed each instrument once (except the VST stuff which I do not build on my main machine) and no memory holes related to the LMMS code were reported. 

Did you play any notes through them as well?

**Comment by michaelgregorius** [comment]
@softrabbit No, I have only added them and removed them.
